### ISSUE TITLE
Need pattern for updating controllers that works with extensions (TPR/Aggregated) 

### ISSUE THREAD

**Issue by erictune** [kubernetes/kubernetes#41677]
Users may also write their own controllers as extensions  -- Aggregated API Servers  (AASes) or Third Party Resources (TPRs).  

Controllers (core or extensions) may contain pod templates. 
Those podTemplates need to be updateable, and that should cause a rollout of the new desired state.

We already support rollouts for deployments by creating a new replicaSet each time we see an update to the controller.

There is [a proposal for how to do this for daemon sets](https://github.com/kubernetes/community/pull/366).  That may eventually use stand-alone pod templates (`/api/v1/podtemplates`).  As noted on that proposal by @janetkuo, the proposed pattern cannot be used by extensions.  

This issue summarizes some problems with extensions that want to have podTemplates.

- **Some Extensions cannot use `observedGeneration`**
  - `templateGeneration` is a field that updates every time the `.spec.podTemplate` of the controller is updated.  It happens in the API server.
  - AAS can support the `observedGeneration`.
  - TPR-based controller can not.
- **Extensions cannot use hashes to compare embedded and standalone pod templates**
  - Deployment computes a `pod-template-hash`.  One issue is that you can GET an object once before a master upgrade, and then GET it after, and it will be different, even though the object was not updated by the user.  This is because new defaulted fields can be added by the new version of the API server code.
  - If an AAS object includes a podTemplate in its controller Spec, then presumably it will have defaulting too.  But that will be vendored from Kubernetes, and so the version of the defaulting in the AAS may be different from the version in the Primary API Server.  
  - TPRs cannot have defaulting.  So they cannot compare their hashes (which are stable) with podTemplate hashes, which are not.
- **Extensions cannot use deep equality to compare embedded and standalone pod templates**
  - Same reasons as above. 

This makes it hard to build controllers that are extensions and have updating and pod templates.

Some things that would solve that problem:

1. If we never used embedded pod templates and just used standalone podTemplates.  So, you have to create at least two objects to make a controller.  But this works poorly with declarative config (apply).
2. If we never apply defaulting or conversion to embedded pod templates. Let that happen at pod creation time.  I like this.  It puts TPR and AAS and core controllers on the same footing.
3. If we figured out a way to generalize the `observedGeneration` pattern to work with TPRs.  

I like 2 but it looks like the most work.  @thockin pointed out that recent changes to defaulting make it difficult to make defaulting not happen in embedded pod templates.

**Comment by erictune** [comment]
@bgrant0607 this is about whether podTemplates play well with API extensions.

**Comment by thockin** [comment]
I agree (2) is correct.  After our chat, I realized I was mixing up
code-gens in my mind. For defaulting I thought the code generator does
something like:

For each top-level API type:
  If the type has a defaulter function, use that
  else
    For each field
      if the field's type has a defaulter function, call it

But that doesn't seem to be quite right.  it's more like:

For each top-level API type:
  If the type has a defaulter function, call that first
  For each field
    if the field's type has a defaulter function, call it

The fact that PodTemplate uses PodSpec, and PodSpec has the defaults
applied means you see this behavior.

we could:

a) use a tag or comment to say "do not apply defaults to this" and put that
on every PodTemplate use-site
b) use a comment on type PodTemplate to say "do not apply defaults to this"
c) track a bit per field of in the internal types indicating "was
defaulted" and when we convert back to versioned, omit those fields.

Maybe other tricks I have not thought of.

I think (2) is reasonable.  In fact, it is so reasonable, Clayton saw fit
to already support it, but it doesn't seem to work as I expect.  That could
probably be fixed in short-ish order.

**Comment by kargakis** [comment]
> Deployment computes a pod-template-hash. One issue is that you can GET an object once before a master upgrade, and then GET it after, and it will be different, even though the object was not updated by the user. This is because new defaulted fields can be added by the new version of the API server code.

Unfortunately, this can happen not only due to defaulting but also net-new fields added in a backwards-compatible fashion as nil pointers produce a new hash different from the previous one.

cc: @kubernetes/sig-api-machinery-misc @kubernetes/sig-apps-misc

**Comment by deads2k** [comment]
> I think (2) is reasonable.  In fact, it is so reasonable, Clayton saw fit
> to already support it, but it doesn't seem to work as I expect.  That could
> probably be fixed in short-ish order.

@smarterclayton

**Comment by smarterclayton** [comment]
I agree with 2 as well.  I know a few people tried it and didn't work as expected, but don't think it's anything more than a logical bug in recursion and loop unrolling (since we have to generate code that walks the defaulters, we need to track different types of nested constructs like loops, maps, and field accessors, and it's entirely possible I missed a flag somewhere).

We've also had cases where people wanted relaxed validation of pod templates - we might actually need that as well because some defaulting is required for validation to pass.

**Comment by thockin** [comment]
What defaulting of PodTemplate is required for validation?  That seems
wrong, on the surface.

WRT default generator - if the generator called per-type functions, rather
than recursing inline, then PodTemplate could explicitly opt-out, and it
would detect that there is no default function for PodTemplate and stop
there.

i.e.

instead of:

```
DefaultsFor Deployment(d) {
    DefaultsForPodSpec(d.template.spec)
}
```

it did

```
DefaultsFor Deployment(d) {
    // DefaultsForPodTemplate(d.template)  // function does not exist, so
not emitted
}
```

**Comment by bgrant0607** [comment]
Some comments:
* I agree we shouldn't / needn't set default values in the pod template
* I hope to do some form of #17333 this year, which would provide un-defaulted configured state more generally
* I think third-party controllers (and probably first-party ones, too) should use standalone pod templates (#170). There are at least 2 approaches that would work well with apply, and we could add generation support to that API. I've also been thinking we should provide a pod monitoring service for third-party controllers (and maybe for first-party ones), as well as subresource support in TPR. 
* The template hash is only supposed to be used to facilitate idempotent creation. We could move to a generation number for that. We should be using strategic merge diff to determine whether there has been a change.
